kotlinx-coroutines-core/common/src/flow/operators/Share.kt

```diff
@@ -359,20 +358,30 @@ public fun <T> MutableSharedFlow<T>.asSharedFlow(): SharedFlow<T> =
 public fun <T> MutableStateFlow<T>.asStateFlow(): StateFlow<T> =
     ReadonlyStateFlow(this)
 
-private class ReadonlySharedFlow<T>(
+private open class ReadonlySharedFlow<T>(
     flow: SharedFlow<T>
 ) : SharedFlow<T> by flow, CancellableFlow<T>, FusibleFlow<T> {
     override fun fuse(context: CoroutineContext, capacity: Int, onBufferOverflow: BufferOverflow) =
         fuseSharedFlow(context, capacity, onBufferOverflow)
 }
 
-private class ReadonlyStateFlow<T>(
+private class ReadonlySharedFlowWithJob<T>(
+    flow: SharedFlow<T>,
+    private val job: Job // keeps a strong reference to the job
+) : ReadonlySharedFlow<T>(flow)
+
+private open class ReadonlyStateFlow<T>(
     flow: StateFlow<T>
 ) : StateFlow<T> by flow, CancellableFlow<T>, FusibleFlow<T> {
     override fun fuse(context: CoroutineContext, capacity: Int, onBufferOverflow: BufferOverflow) =
         fuseStateFlow(context, capacity, onBufferOverflow)
 }
 
+private class ReadonlyStateFlowWithJob<T>(
+    flow: StateFlow<T>,
+    private val job: Job // keeps a strong reference to the job
```

Comment:
Please add suppress in order to avoid warnings and elaborate a bit on the purpose of this in the comment for us from the future :)

Reply:
Done